### PR TITLE
[move-stdlib] Introduce efficient OrderedMap implementation

### PR DESCRIPTION
**Pull request by igor-aptos**
## Description
Implement efficient OrderedMap implementation in move-stdlib, to replace/deprecate SimpleMap. It is an in-memory datastructure - i.e. doesn't store anything in separate resources.

Implementation is a SortedVectorMap, but given the efficiency improvements of vector operations with memcpy, it's performance characteristics are extremely good. 

Running different sizes tests (these are [full results](https://gist.github.com/igor-aptos/815f06754868db8ecedf1714e953af33)), we get milliseconds needed for 100 inserts and 100 removals on maps of different sizes:

```
num elements  SimpleMap       OrderedMap::SortedVectorMap
10            6.4             6.8
100           25.4            10.3
1000          673.9           12.8
10000         too long        19.5
```

Basically achieving log(n) performance, with very simple implementation.
 

## How Has This Been Tested?
provided unit tests

## Type of Change
- [x] New feature

## Which Components or Systems Does This Change Impact?
- [x] Aptos Framework

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+/// Additionally, it provides a way to lookup and iterate over sorted keys, making range query
+/// take O(log N + R) time (where R is number of elements in the range).
+///
+/// Most methods operate with OrderedMap being `self`.
+/// All methods that start with iter_*, operate with Iterator being `self`.
+///
+/// Uses cmp::compare for ordering, which compares primitive types natively, and uses common
+/// lexicographical sorting for complex types.
+module aptos_std::ordered_map {
+    // friend aptos_std::big_ordered_map;
```
Comment (lightmark) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:20]: ？

Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:20]: will be uncommented in the PR with big_ordered_map

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+    /// Individual entry holding (key, value) pair
+    struct Entry<K, V> has drop, copy, store {
+        key: K,
+        value: V,
+    }
+
+    /// The OrderedMap datastructure.
+    enum OrderedMap<K, V> has drop, copy, store {
+        /// sorted-vector based implementation of OrderedMap
+        SortedVectorMap {
```
Comment (lightmark) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:47]: [cool](https://app.graphite.dev//settings/meme-library?org=aptos-labs)

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+        }
+    }
+
+    /// Remove a key/value pair from the map.
+    /// Aborts with EKEY_NOT_FOUND if `key` doesn't exist.
+    public fun remove<K: drop, V>(self: &mut OrderedMap<K, V>, key: &K): V {
+        let len = self.entries.length();
+        let index = binary_search(key, &self.entries, 0, len);
+        assert!(index < len, error::invalid_argument(EKEY_NOT_FOUND));
+        let Entry { key: old_key, value } = self.entries.remove(index);
```
Comment (lightmark) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:116]: vector does not keep the order here? IIRC, it `swap` and `pop_back`. 
This is critical.

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+    /// Whether map is empty.
+    public fun is_empty<K, V>(self: &OrderedMap<K, V>): bool {
+        self.entries.is_empty()
+    }
+
+    /// Add a key/value pair to the map.
+    /// Aborts with EKEY_ALREADY_EXISTS if key already exist.
+    public fun add<K, V>(self: &mut OrderedMap<K, V>, key: K, value: V) {
+        let len = self.entries.length();
+        let index = binary_search(&key, &self.entries, 0, len);
```
Comment (lightmark) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:85]: we should introduce iterator to vector too so you don't need to know the length anymore for assertion.

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+            return;
+        };
+
+        // Optimization: if all elements in `other` are larger than all elements in `self`, we can just move them over.
+        if (cmp::compare(&self.entries.borrow(self.entries.length() - 1).key, &other_entries.borrow(0).key).is_less_than()) {
+            self.entries.append(other_entries);
+            return;
+        };
+
+        // TODO: can be implemented more efficiently, as we know both maps are already sorted.
```
Comment (lightmark) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:169]: https://leetcode.com/problems/merge-two-sorted-lists/

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+    // ========= Section with inline for-loop methods =======
+
+    /// Apply the function to a reference of each key-value pair in the table.
+    public inline fun for_each_ref<K, V>(self: &OrderedMap<K, V>, f: |&K, &V|) {
+        let iter = self.new_begin_iter();
+        while (!iter.iter_is_end()) {
+            f(iter.iter_borrow_key(self), iter.iter_borrow(self));
+            iter = iter.iter_next(self);
+        }
+        // vector::for_each_ref(
```
Comment (lightmark) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:349]: what's wrong here?

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+
+    inline fun new_iter(index: u64): Iterator {
+        Iterator::Position {
+            index: index,
+        }
+    }
+
+    // return index containing the key, or insert position.
+    // I.e. index of first element that has key larger or equal to the passed `key` argument.
+    fun binary_search<K, V>(key: &K, entries: &vector<Entry<K, V>>, start: u64, end: u64): u64 {
```
Comment (lightmark) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:383]: Can we make a standard `Entry` and implement binary search only once?

**Review by lightmark — APPROVED** [review]
amazing work! Thanks @igor-aptos . You are the hero!

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move (hunk 3 above)
Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:116]: `remove` keeps the order. `swap_remove` doesn't

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move (hunk 4 above)
Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:85]: this seems fine to me

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+
+    /// Returns whether map contains a given key.
+    public fun contains<K, V>(self: &OrderedMap<K, V>, key: &K): bool {
+        !self.find(key).iter_is_end()
+    }
+
+    /// Changes the key, with keeping the same value attached to it
+    /// Aborts with EKEY_NOT_FOUND if `old_key` doesn't exist.
+    /// Aborts with ENEW_KEY_NOT_IN_ORDER if `new_key` doesn't keep the order `old_key` was in.
+    public(friend) fun replace_key_inplace<K: drop, V>(self: &mut OrderedMap<K, V>, old_key: &K, new_key: K) {
```
[the remark this answers is not in the document]
Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:129]: seems not generally useful, but in BTreeMap keys here represent max_key of a child subtree, and so I need to modify it in place.

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move (hunk 5 above)
Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:169]: yeah, but issue is here we have vectors, not lists, and so how to do it in-place, without moving too many elements.

Easiest is to can either create a new vector, populate, and then swap (I'll probably just do that)

if "vector" here had capacity, I could start from the end, without temporary storage, but I don't have a way to create empty values here (i.e. have this work without value being drop+copy)

**Review by msmouse — COMMENTED** [review]
I feel it's the best if someone works more on the platform approves..

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 9]
```diff
@@ -0,0 +1,786 @@
...
+
+        reverse_result.reverse_append(other_entries);
+        self.entries.reverse_append(reverse_result);
+    }
+
+    /// Splits the collection into two, such to leave `self` with `at` number of elements.
+    /// Returns a newly allocated map containing the elements in the range [at, len).
+    /// After the call, the original map will be left containing the elements [0, at)
+    /// with its previous capacity unchanged.
+    public fun trim<K, V>(self: &mut OrderedMap<K, V>, at: u64): OrderedMap<K, V> {
```
Comment (grao1991) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:230]: should we call it `split_off`?

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 10]
```diff
@@ -0,0 +1,786 @@
...
+
+    inline fun new_iter(index: u64): Iterator {
+        Iterator::Position {
+            index: index,
+        }
+    }
+
+    // return index containing the key, or insert position.
+    // I.e. index of first element that has key larger or equal to the passed `key` argument.
+    fun binary_search<K, V>(key: &K, entries: &vector<Entry<K, V>>, start: u64, end: u64): u64 {
```
Comment (grao1991) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:478]: Should we just call this function `lower_bound`?

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 11]
```diff
@@ -0,0 +1,786 @@
...
+        let l = start;
+        let r = end;
+        while (l != r) {
+            let mid = l + (r - l) / 2;
+            let comparison = cmp::compare(&entries.borrow(mid).key, key);
+            // TODO: check why this short-circuiting actually performs worse
+            // if (comparison.is_equal()) {
+            //     // there can only be one equal value, so end the search.
+            //     return mid;
+            // } else
```
[the remark this answers is not in the document]
Comment (grao1991) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:488]: most std lib in c++ doesn't do short-circuiting.

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 12]
```diff
@@ -0,0 +1,786 @@
...
+        }
+    }
+
+    // return index containing the key, or insert position.
+    // I.e. index of first element that has key larger or equal to the passed `key` argument.
+    fun binary_search<K, V>(key: &K, entries: &vector<Entry<K, V>>, start: u64, end: u64): u64 {
+        let l = start;
+        let r = end;
+        while (l != r) {
+            let mid = l + (r - l) / 2;
```
Comment (grao1991) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:482]: would `>> 1` be faster here or they are the same?

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 13]
```diff
@@ -0,0 +1,786 @@
...
+        Iterator::Position {
+            index: index,
+        }
+    }
+
+    // return index containing the key, or insert position.
+    // I.e. index of first element that has key larger or equal to the passed `key` argument.
+    fun binary_search<K, V>(key: &K, entries: &vector<Entry<K, V>>, start: u64, end: u64): u64 {
+        let l = start;
+        let r = end;
```
Comment (grao1991) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:480]: use l and len (r - l) here, and do len/2 every time can probably save a little bit.

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move (hunk 9 above)
Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:230]: I would, but we already have vector::trim that does split_off, so for consistency better to use the same

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 14]
```diff
@@ -0,0 +1,786 @@
...
+    }
+
+    /// For maps that cannot be dropped this is a utility to destroy them
+    /// using lambdas to destroy the individual keys and values.
+    public inline fun destroy<K, V>(
+        self: OrderedMap<K, V>,
+        dk: |K|,
+        dv: |V|
+    ) {
+        let (keys, values) = to_vec_pair(self);
```
[the remark this answers is not in the document]
Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:432]: move doesn't support public types yet :)

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move (hunk 13 above)
Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:480]: don't get what you mean, can you clarify? this is also your implementation I copied here, by the way :P

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move (hunk 10 above)
Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:478]: we already have a lower_bound which returns an iterator.

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move (hunk 13 above)
Comment (grao1991) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:480]: Something like this: https://github.com/gcc-mirror/gcc/blob/d9375e490072d1aae73a93949aa158fcd2a27018/libstdc%2B%2B-v3/include/bits/stl_algobase.h#L1023

Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:491]: This seems to only make sense where you can have pointers and do pointer arithmetic. Here since you need to access the vector, and always produce index from the start, you need to compute the absolute index anyways - so I don't think that would be faster.

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 15]
```diff
@@ -0,0 +1,786 @@
...
+        if (index > 0) {
+            assert!(cmp::compare(&self.entries[index - 1].key, &new_key).is_lt(), error::invalid_argument(ENEW_KEY_NOT_IN_ORDER))
+        };
+
+        if (index + 1 < len) {
+            assert!(cmp::compare(&new_key, &self.entries[index + 1].key).is_lt(), error::invalid_argument(ENEW_KEY_NOT_IN_ORDER))
+        };
+
+        let entry = self.entries.borrow_mut(index);
+        assert!(old_key == &entry.key, error::invalid_argument(EKEY_NOT_FOUND));
```
[the remark this answers is not in the document]
Comment (msmouse) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:165]: (I meant I would assert fouund before asserting the order)

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move [hunk 16]
```diff
@@ -0,0 +1,786 @@
...
+    /// Add multiple key/value pairs to the map. The keys must not already exist.
+    /// Aborts with EKEY_ALREADY_EXISTS if key already exist, or duplicate keys are passed in.
+    public fun add_all<K, V>(self: &mut OrderedMap<K, V>, keys: vector<K>, values: vector<V>) {
+        vector::zip(keys, values, |key, value| {
+            add(self, key, value);
+        });
+    }
+
+    /// Takes all elements from `other` and adds them to `self`.
+    /// Aborts with EKEY_ALREADY_EXISTS if `other` has a key already present in `self`.
```
[the remark this answers is not in the document]
Comment (msmouse) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:188]: append_unique or append_disjoint? I think overwriting should be the default (as in rust btreemap, for example)
 no big deal since it's well documented.

**Review by msmouse — APPROVED** [review]

File: aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move (hunk 16 above)
[the remark this answers is not in the document]
Comment (igor-aptos) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move:188]: side-note - the reasoning for having this variant (and not being able to have a boolean to choose), or just assert on the size of the output, is that overwriting requires "drop".
